src/librustc_infer/traits/wf.rs

```diff
@@ -322,6 +323,8 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
             let implied_obligations = traits::elaborate_obligations(tcx, obligations.clone());
             let implied_obligations = implied_obligations.map(|obligation| {
                 let mut cause = cause.clone();
+                cause.code =
+                    traits::ObligationCauseCode::DerivedObligation(Rc::new(obligation.cause.code));
```

Comment:
Wait, can't you use `trait_ref` to make an `ImplDerivedObligation`? For some reason I thought this was a general WF thing but it's more specific than that.

Reply:
(or `BuiltinDerivedObligation`)

The error messages might be weird either way, but it shouldn't be hard to try and I'm curious what the test diff looks like.

Reply:
Last commit uses `ImplDerivedObligation`, it actually works better than I expected.